### PR TITLE
navgen: properly increase cell height when needed

### PR DESCRIPTION
**Pull request by illwieckz**
Ported from daemonmap:

- https://github.com/DaemonEngine/daemonmap/pull/13

### CODE REVIEW

**Comment by illwieckz** [comment]
It generates anthill navmeshes:

```
==== Bot Navigation Initialization ====
Warn: Map geometry is too tall for specified cell height. Increasing cell height to compensate. This may cause a less accurate navmesh.
Generating navmesh for level0
[bot] Bot#1 is connecting…
[bot] Bot#1 entered the game
[bot] Bot#2 is connecting…
[bot] Bot#2 entered the game
[bot] Bot#3 is connecting…
[bot] Bot#3 entered the game
[bot] Bot#4 is connecting…
[bot] Bot#4 entered the game
[bot] Bot#5 is connecting…
[bot] Bot#5 entered the game
[bot] Bot#6 is connecting…
[bot] Bot#6 entered the game
Server: generating bot navigation mesh for level0... 0%
Server: generating bot navigation mesh for level0... 53%
Server: generating bot navigation mesh for level0... 80%
Generating navmesh for level1
Server: generating bot navigation mesh for level1... 50%
Server: generating bot navigation mesh for level1... 76%
Generating navmesh for level2
Server: generating bot navigation mesh for level2... 68%
Generating navmesh for level2upg
Server: generating bot navigation mesh for level2upg... 77%
Generating navmesh for level3
Generating navmesh for level3upg
Server: generating bot navigation mesh for level3upg... 32%
Generating navmesh for level4
Server: generating bot navigation mesh for level4... 70%
Generating navmesh for human_naked
Server: generating bot navigation mesh for human_naked... 50%
Server: generating bot navigation mesh for human_naked... 77%
Generating navmesh for human_bsuit
Server: generating bot navigation mesh for human_bsuit... 48%
Server: generating bot navigation mesh for human_bsuit... 73%
==== Bot Navigation Initialization ====
 loading navigation mesh file 'maps/anthill-level0.navMesh'...
Warn: Loading navmesh for level0 failed: Navgen config changed
Warn: Cannot load navigation mesh file for level0
Bot navmesh generation failed!
[bot] Bot#1 disconnected
[bot] Bot#2 disconnected
[bot] Bot#3 disconnected
[bot] Bot#4 disconnected
[bot] Bot#5 disconnected
[bot] Bot#6 disconnected
]/navedit enable level0
Warn: Couldn't load navmeshes
]/bot fill 3
Warn: Navmesh initialization previously failed, doing nothing
Navigation mesh files unavailable for this map

]/devmap anthill
]/bot fill 3
==== Bot Navigation Initialization ====
 loading navigation mesh file 'maps/anthill-level0.navMesh'...
Warn: Loading navmesh for level0 failed: Navgen config changed
 loading navigation mesh file 'maps/anthill-level1.navMesh'...
Warn: Loading navmesh for level1 failed: Navgen config changed
 loading navigation mesh file 'maps/anthill-level2.navMesh'...
Warn: Loading navmesh for level2 failed: Navgen config changed
 loading navigation mesh file 'maps/anthill-level2upg.navMesh'...
Warn: Loading navmesh for level2upg failed: Navgen config changed
 loading navigation mesh file 'maps/anthill-level3.navMesh'...
Warn: Loading navmesh for level3 failed: Navgen config changed
 loading navigation mesh file 'maps/anthill-level3upg.navMesh'...
Warn: Loading navmesh for level3upg failed: Navgen config changed
 loading navigation mesh file 'maps/anthill-level4.navMesh'...
Warn: Loading navmesh for level4 failed: Navgen config changed
 loading navigation mesh file 'maps/anthill-human_naked.navMesh'...
Warn: Loading navmesh for human_naked failed: Navgen config changed
 loading navigation mesh file 'maps/anthill-human_bsuit.navMesh'...
Warn: Loading navmesh for human_bsuit failed: Navgen config changed
Warn: Map geometry is too tall for specified cell height. Increasing cell height to compensate. This may cause a less accurate navmesh.
Generating navmesh for level0
[bot] Bot#1 is connecting…
[bot] Bot#1 entered the game
[bot] Bot#2 is connecting…
[bot] Bot#2 entered the game
[bot] Bot#3 is connecting…
[bot] Bot#3 entered the game
[bot] Bot#4 is connecting…
[bot] Bot#4 entered the game
[bot] Bot#5 is connecting…
[bot] Bot#5 entered the game
[bot] Bot#6 is connecting…
[bot] Bot#6 entered the game
Server: generating bot navigation mesh for level0... 0%
Server: generating bot navigation mesh for level0... 53%
Server: generating bot navigation mesh for level0... 81%
Generating navmesh for level1
illwieckz [granger] joined the humans
[bot] Bot#6 disconnected
Server: generating bot navigation mesh for level1... 49%
Server: generating bot navigation mesh for level1... 76%
Generating navmesh for level2
Server: generating bot navigation mesh for level2... 70%
Generating navmesh for level2upg
Server: generating bot navigation mesh for level2upg... 78%
Generating navmesh for level3
Generating navmesh for level3upg
Server: generating bot navigation mesh for level3upg... 36%
Generating navmesh for level4
Server: generating bot navigation mesh for level4... 76%
Generating navmesh for human_naked
Server: generating bot navigation mesh for human_naked... 50%
Server: generating bot navigation mesh for human_naked... 78%
Generating navmesh for human_bsuit
Server: generating bot navigation mesh for human_bsuit... 47%
Server: generating bot navigation mesh for human_bsuit... 74%
==== Bot Navigation Initialization ====
 loading navigation mesh file 'maps/anthill-level0.navMesh'...
Warn: Loading navmesh for level0 failed: Navgen config changed
Warn: Cannot load navigation mesh file for level0
Bot navmesh generation failed!
[bot] Bot#1 disconnected
[bot] Bot#2 disconnected
[bot] Bot#3 disconnected
[bot] Bot#4 disconnected
[bot] Bot#5 disconnected
```

But they are always invalidated, maybe navmesh invalidation is always true on dpkdir?

**Comment by illwieckz** [comment]
Maybe the reason why the navmesh is always invalidated is that the algorithm stores the cellheight after generation but then compares against default cellheight.

**Comment by illwieckz** [comment]
Now it works, I excluded cellHeight comparison when validating navMeshes.

**Comment by illwieckz** [comment]
```
]/devmap anthill
]/bot fill 3
==== Bot Navigation Initialization ====
Warn: Map geometry is too tall for specified cell height. Increasing cell height to compensate. This may cause a less accurate navmesh.
Generating navmesh for level0
[bot] Bot#1 is connecting…
[bot] Bot#1 entered the game
[bot] Bot#2 is connecting…
[bot] Bot#2 entered the game
[bot] Bot#3 is connecting…
[bot] Bot#3 entered the game
[bot] Bot#4 is connecting…
[bot] Bot#4 entered the game
[bot] Bot#5 is connecting…
[bot] Bot#5 entered the game
[bot] Bot#6 is connecting…
[bot] Bot#6 entered the game
Server: generating bot navigation mesh for level0... 0%
Server: generating bot navigation mesh for level0... 53%
Server: generating bot navigation mesh for level0... 80%
Generating navmesh for level1
Server: generating bot navigation mesh for level1... 49%
Server: generating bot navigation mesh for level1... 76%
Generating navmesh for level2
Server: generating bot navigation mesh for level2... 68%
Generating navmesh for level2upg
Server: generating bot navigation mesh for level2upg... 76%
Generating navmesh for level3
Generating navmesh for level3upg
Server: generating bot navigation mesh for level3upg... 23%
Generating navmesh for level4
Server: generating bot navigation mesh for level4... 64%
Generating navmesh for human_naked
Server: generating bot navigation mesh for human_naked... 47%
Server: generating bot navigation mesh for human_naked... 73%
Generating navmesh for human_bsuit
Server: generating bot navigation mesh for human_bsuit... 43%
Server: generating bot navigation mesh for human_bsuit... 68%
==== Bot Navigation Initialization ====
 loading navigation mesh file 'maps/anthill-level0.navMesh'...
 loading navigation mesh file 'maps/anthill-level1.navMesh'...
 loading navigation mesh file 'maps/anthill-level2.navMesh'...
 loading navigation mesh file 'maps/anthill-level2upg.navMesh'...
 loading navigation mesh file 'maps/anthill-level3.navMesh'...
 loading navigation mesh file 'maps/anthill-level3upg.navMesh'...
 loading navigation mesh file 'maps/anthill-level4.navMesh'...
 loading navigation mesh file 'maps/anthill-human_naked.navMesh'...
 loading navigation mesh file 'maps/anthill-human_bsuit.navMesh'...
 ```

**Comment by illwieckz** [comment]
unvanquished_2023-02-02_031511_000

**Comment by slipher** [comment]
1. The "algorithm" where the map height is repeatedly divided by something in a loop makes no sense. We shouldn't bring it back.
2. The point of NavgenConfig is to put parameters that MUST be the same, else the cache is invalidated. We should keep the *requested* height in there and then the actual height can be stored somewhere else.

If fixing this stuff would be a lot of trouble I can make my own version.

**Comment by illwieckz** [comment]
So, what if we just do `cellHeight = height / RC_SPAN_MAX_HEIGHT` if `height / cellHeight > RC_SPAN_MAX_HEIGHT`, like I first did on dæmonmap?

**Comment by illwieckz** [comment]
> The point of NavgenConfig is to put parameters that MUST be the same, else the cache is invalidated. We should keep the requested height in there and then the actual height can be stored somewhere else.

I agree with that but I don't know how to do it without breaking the navmesh file format.

**Comment by necessarily-equal** [comment]
> I agree with that but I don't know how to do it without breaking the navmesh file format.

I think the navmesh file format was broken already, so it doesn't change much at this point

**Comment by illwieckz** [comment]
Well, I can also break the format and bump the navmesh version number…

**Comment by illwieckz** [comment]
I made a simpler patch to just not store the computed cellHeight…

**Comment by necessarily-equal** [comment]
This is simpler, but one may wonder if this could be bug prone: it would be easy to confuse both cellHeight and config_.cellHeight around, and the comment isn't explicit about that hiccup either.

File: src/shared/navgen/navgen.h [hunk 1]
```diff
@@ -140,6 +140,8 @@ class NavmeshGenerator {
 
 	UnvContext recastContext_;
 	NavgenConfig config_;
+	// Custom computed config for the map
+	float cellHeight;
```
Comment (slipher) [inline src/shared/navgen/navgen.h:144]: Use naming convention with underscore

File: src/shared/navgen/nav.cpp [hunk 2]
```diff
@@ -1061,9 +1061,17 @@ void NavmeshGenerator::Init(Str::StringRef mapName)
 	LoadGeometry();
 	if ( initStatus_.code != NavgenStatus::OK ) return;
 
+	cellHeight = config_.cellHeight;
```
Comment (slipher) [inline src/shared/navgen/nav.cpp:1064]: Let's name the one in config to `requestedCellHeight` to reduce chances for confusion

**Review by slipher — COMMENTED** [review]
Don't forget to change NAVMESHSET_VERSION. Use 8 since I have an open PR with 7

**Comment by illwieckz** [comment]
> Don't forget to change NAVMESHSET_VERSION. Use 8 since I have an open PR with 7

This doesn't break the format since I don't store the requested cellHeight… and I doubt any existing navmesh stored a non-default cellHeight…

**Comment by slipher** [comment]
Yeah the sole use of changing the version would be to invalidate cached failures.

File: src/shared/navgen/nav.cpp (hunk 2 above)
Comment (illwieckz) [reply src/shared/navgen/nav.cpp]: done

File: src/shared/navgen/navgen.h (hunk 1 above)
Comment (illwieckz) [reply src/shared/navgen/navgen.h]: done

**Comment by illwieckz** [comment]
> Yeah the sole use of changing the version would be to invalidate cached failures.

Ah, that's a good point!

**Comment by illwieckz** [comment]
I rebased and squashed.

**Comment by slipher** [comment]
LGTM

**Pull request merged**